### PR TITLE
Mimic SPM's logic for obtaining the minimum deployment target

### PR DESCRIPTION
**Pull request by pepibumur**
Fixes https://github.com/tuist/tuist/issues/3633
Related https://github.com/tuist/tuist/pull/3602

### Short description 📝
As it was reported [here](https://github.com/tuist/tuist/issues/3633) the logic that we use for determining the minimum deployment target for macOS doesn't return the right version in the case of macOS. As a consequence, developing Tuist through Tuist generates an Xcode project that fails to compile.

This PR fixes is by mimicking [SPM's output parsing](https://github.com/apple/swift-package-manager/blob/fb989898f76e0c694a4547ba7ffc8b0e40199545/Sources/Workspace/DefaultPluginScriptRunner.swift#L160). I've taken the opportunity to extract the logic into its own component that is easily reusable and testable.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.
- [ ] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 
 - Fix handling of `TUIST_CONFIG_COLOURED_OUTPUT` environment variable [#3631](https://github.com/tuist/tuist/pull/3631) by [@danyf90](https://github.com/danyf90)
 - Fix `tuist dump config` no longer requires to be executed inside the `Tuist` folder [#3647](https://github.com/tuist/tuist/pull/3647) by [@danyf90](https://github.com/danyf90)
+- Calculation of the macOS minimum deployment target for Swift Packages that don't specify it [#3639](https://github.com/tuist/tuist/pull/3639) by [@pepibumur](https://github.com/pepibumur)
```
Comment (danyf90) [inline CHANGELOG.md:23]: I think we don't need a changelog entry, as we already have the one introducing the "feature" (the first in Changes)

File: Tuist/Dependencies/Lockfiles/Package.resolved [hunk 2]
```diff
@@ -1,259 +0,0 @@
-{
```
Comment (danyf90) [inline Tuist/Dependencies/Lockfiles/Package.resolved]: This shouldn't be deleted, should it?

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift [hunk 3]
```diff
@@ -788,6 +774,27 @@ extension TuistGraph.Platform {
...
+        }
+    }
+}
+
+extension TuistGraph.DeploymentDevice {
+    fileprivate var descriptionDeploymentDevice: ProjectDescription.DeploymentDevice {
+        return .init(rawValue: rawValue)
+    }
+}
+
```
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:797]: I think these are not used anymore

Comment (pepibumur) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:797]: Right. I'll remove them.

File: CHANGELOG.md [hunk 4]
```diff
@@ -20,7 +20,6 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 
 - Fix handling of `TUIST_CONFIG_COLOURED_OUTPUT` environment variable [#3631](https://github.com/tuist/tuist/pull/3631) by [@danyf90](https://github.com/danyf90)
 - Fix `tuist dump config` no longer requires to be executed inside the `Tuist` folder [#3647](https://github.com/tuist/tuist/pull/3647) by [@danyf90](https://github.com/danyf90)
-
```
Comment (danyf90) [inline CHANGELOG.md:23]: We should keep the blank line :)

**Review by danyf90 — APPROVED** [review]

**Comment by fortmarek** [comment]
Does working on tuist with tuist work for you? I still keep getting an error with the minimum deployment target, @pepibumur @danyf90

**Comment by danyf90** [comment]
> Does working on tuist with tuist work for you? I still keep getting an error with the minimum deployment target, @pepibumur @danyf90

I have just tried locally and it seems to work fine 🤔

**Pull request merged**